### PR TITLE
Remove Underscore [1 of 2]

### PR DESCRIPTION
**Pull request by daiyiwhale**
Reviewer: @mlunoe

### CODE REVIEW

**Comment by daiyiwhale** [comment]
Mission Control to Velocity 📢

File: src/js/structs/StateSummary.js [hunk 1]
```diff
@@ -53,7 +61,9 @@ module.exports = class StateSummary {
   }
 
   getActiveSlaves() {
-    return _.where(this.snapshot.slaves, {active: true});
+    return this.snapshot.slaves.filter(function (slave) {
```
Comment (kennyt) [inline src/js/structs/StateSummary.js]: there is a slight difference here, where previously we would only get things that had `active === true`, now we get things that are truthy and not exactly true. probably won't matter, though

File: src/js/structs/StateSummary.js [hunk 2]
```diff
@@ -30,17 +28,27 @@ module.exports = class StateSummary {
...
     }
     this.metadata.date = options.date || Date.now();
+
+    let slaves = this.snapshot.slaves || [];
     // Store computed data – this is something we may not need to store
     this.metadata.slaveTotalResources = MesosSummaryUtil.sumResources(
       // We may only want to get the active slaves...
-      _.pluck(this.snapshot.slaves, 'resources')
+      slaves.map(function (slave) {
+        return slave.resources || null;
```
Comment (kennyt) [inline src/js/structs/StateSummary.js]: we don't really need the `|| null` for all of these, because the original `_.pluck` will return `undefined` for them anyways...

plus, `slave.resources` might be something like an empty string, 0, etc. so I elect to just return `slave.resources` straight up (same for all of the pluck replaces)

File: src/js/pages/system/OrganizationTab.js [hunk 3]
```diff
@@ -1,5 +1,3 @@
-import _ from 'underscore';
-import classNames from 'classnames';
```
Comment (kennyt) [inline src/js/pages/system/OrganizationTab.js]: why is it that we remove `classNames`? has it never been used?

File: src/js/components/modals/ActionsModal.js [hunk 4]
```diff
@@ -108,18 +108,20 @@ class ActionsModal extends mixin(StoreMixin) {
       selectedItemsString = selectedItems[0].description;
     } else {
       // Truncate list of selected user/groups for ease of reading
-      let selectedItemsShown = _.first(selectedItems, ITEMS_DISPLAYED + 1);
+      let selectedItemsShown = selectedItems.slice(0, ITEMS_DISPLAYED);
```
Comment (kennyt) [inline src/js/components/modals/ActionsModal.js]: are you sure you don't need the `+1` anymore? i'm doing it in the console, and first/slice returns the same number of items if the 2nd arg is the same

**Comment by kennyt** [comment]
i get this on opening side panel

 px

File: src/js/structs/StateSummary.js (hunk 1 above)
Comment (daiyiwhale) [inline src/js/structs/StateSummary.js]: It didn't make sense to me that `.active` would be anything but a boolean but it can be changed back to explicit `slave.active === true` if you think it's important?

File: src/js/pages/system/OrganizationTab.js (hunk 3 above)
Comment (daiyiwhale) [inline src/js/pages/system/OrganizationTab.js]: oops, that should still be there. good catch

File: src/js/components/modals/ActionsModal.js (hunk 4 above)
Comment (daiyiwhale) [inline src/js/components/modals/ActionsModal.js]: 👍

File: src/js/structs/StateSummary.js (hunk 2 above)
Comment (daiyiwhale) [inline src/js/structs/StateSummary.js]: good point, changed back!

File: src/js/structs/StateSummary.js (hunk 1 above)
Comment (kennyt) [inline src/js/structs/StateSummary.js]: yes please. who knows what can happen in the future with the API

File: src/js/structs/NodesList.js [hunk 5]
```diff
@@ -58,14 +56,18 @@ module.exports = class NodesList extends List {
 
   sumUsedResources() {
     let services = this.getItems();
-    let resourcesList = _.pluck(services, 'used_resources');
+    let resourcesList = services.map(function (service) {
+      return service.used_resources || null;
```
Comment (kennyt) [inline src/js/structs/NodesList.js]: still got some `|| null`s here

File: src/js/structs/ServicesList.js [hunk 6]
```diff
@@ -24,7 +22,7 @@ module.exports = class ServicesList extends List {
 
     if (filters) {
       if (filters.ids) {
-        services = _.filter(services, function (service) {
+        services = services.filter(function (service) {
           return this.ids.indexOf(service.id) !== -1;
         }, {ids: filters.ids});
```
Comment (kennyt) [inline src/js/structs/ServicesList.js]: i think we might need to make some more changes here. here we pass `{ids: filters.ids}` as the 3rd argument to make it the context. the native `filter` doesn't have that

File: src/js/structs/StateSummary.js [hunk 7]
```diff
@@ -30,17 +28,27 @@ module.exports = class StateSummary {
       this.metadata.successfulSnapshot = options.successful;
     }
     this.metadata.date = options.date || Date.now();
+
+    let slaves = this.snapshot.slaves || [];
```
Comment (kennyt) [inline src/js/structs/StateSummary.js]: don't we already do a `|| []` up above? why do we need to do it again?

File: src/js/utils/Util.js [hunk 8]
```diff
@@ -1,5 +1,33 @@
...
+   * @param  {Object} object        Object to copy
+   * @param  {Array} blacklistKeys  Keys to not copy over
+   * @return {Object}               Copy of object without blacklisted keys
+   */
+  omit: function (object, blacklistKeys) {
+    return Object.keys(object).reduce(function (newObject, key) {
+      if (blacklistKeys.indexOf(key) === -1) {
+        newObject[key] = object[key];
+      }
+      return newObject;
```
Comment (kennyt) [inline src/js/utils/Util.js]: can we put a newline before return?

File: src/js/structs/ServicesList.js (hunk 6 above)
Comment (daiyiwhale) [inline src/js/structs/ServicesList.js]: oh hm, why was it necessary to make `{ids: filters.ids}` the context? can we call `filters.ids` directly (`filters.ids.indexOf(service.id) !== -1`)?

Comment (kennyt) [inline src/js/structs/ServicesList.js]: yeah i think we can. i'm not too sure why it was implemented like that in the beginning

Comment (daiyiwhale) [inline src/js/structs/ServicesList.js]: weird o_O

Comment (kennyt) [inline src/js/structs/ServicesList.js]: ping @rcorral you committed this change before :)

**Comment by kennyt** [comment]
what is the 2nd part of this PR going to do?

**Comment by daiyiwhale** [comment]
Same thing but removes more underscores. It was pretty overwhelming to look at all at once

**Comment by kennyt** [comment]
👍

File: src/js/structs/ServicesList.js [hunk 9]
```diff
@@ -45,7 +43,9 @@ module.exports = class ServicesList extends List {
 
   sumUsedResources() {
     let services = this.getItems();
-    let resourcesList = _.pluck(services, 'used_resources');
+    let resourcesList = services.map(function (service) {
+      return service.used_resources || null;
```
Comment (kennyt) [inline src/js/structs/ServicesList.js]: one more `|| null`

Comment (daiyiwhale) [inline src/js/structs/ServicesList.js]: 👍

File: src/js/stores/NodeHealthStore.js [hunk 10]
```diff
@@ -94,7 +93,7 @@ const NodeHealthStore = Store.createStore({
   },
 
   shouldPoll: function () {
-    return !_.isEmpty(this.listeners(HEALTH_NODES_CHANGE));
+    return (this.listeners(HEALTH_NODES_CHANGE).length === 0);
```
Comment (kennyt) [inline src/js/stores/NodeHealthStore.js]: wait, isn't the original code checking if its *not* empty, and this checks if it *is* empty?

File: src/js/stores/UnitHealthStore.js [hunk 11]
```diff
@@ -94,7 +93,7 @@ const UnitHealthStore = Store.createStore({
   },
 
   shouldPoll: function () {
-    return !_.isEmpty(this.listeners(HEALTH_UNITS_CHANGE));
+    return (this.listeners(HEALTH_UNITS_CHANGE).length === 0);
```
Comment (kennyt) [inline src/js/stores/UnitHealthStore.js]: same here. the original code checks if it's *not* empty, this checks if it *is* empty.

File: src/js/structs/NodesList.js [hunk 12]
```diff
@@ -27,7 +25,7 @@ module.exports = class NodesList extends List {
 
       // Marathon API
       if (filters.ids) {
-        hosts = _.filter(hosts, function (host) {
+        hosts = hosts.filter(function (host) {
           return this.ids.indexOf(host.id) !== -1;
         }, {ids: filters.ids});
```
Comment (kennyt) [inline src/js/structs/NodesList.js]: we have a 3rd argument here again. native filter doesn't have the 3rd argument

File: src/js/stores/__tests__/NodeHealthStore-test.js [hunk 13]
```diff
@@ -65,14 +65,14 @@ describe('NodeHealthStore', function () {
...
-      var mockedFn = jest.genMockFunction();
+      var mockedFn = jasmine.createSpy();
       NodeHealthStore.addChangeListener(EventTypes.HEALTH_NODES_CHANGE, mockedFn);
       AppDispatcher.handleServerAction({
         type: ActionTypes.REQUEST_HEALTH_NODES_SUCCESS,
         data: []
       });
 
-      expect(mockedFn.mock.calls.length).toEqual(2);
+      expect(mockedFn.calls.count()).toEqual(1);
```
Comment (kennyt) [inline src/js/stores/__tests__/NodeHealthStore-test.js]: why would the number of calls drop from 2 to 1? I suspect the test failing was due to the fact that `shouldPoll` is implemented incorrectly and this should go back to `2`, right?

File: src/js/stores/__tests__/UnitHealthStore-test.js [hunk 14]
```diff
@@ -66,14 +66,14 @@ describe('UnitHealthStore', function () {
...
-      var mockedFn = jest.genMockFunction();
+      var mockedFn = jasmine.createSpy();
       UnitHealthStore.addChangeListener(EventTypes.HEALTH_UNITS_CHANGE, mockedFn);
       AppDispatcher.handleServerAction({
         type: ActionTypes.REQUEST_HEALTH_UNITS_SUCCESS,
         data: []
       });
 
-      expect(mockedFn.mock.calls.length).toEqual(2);
+      expect(mockedFn.calls.count()).toEqual(1);
```
Comment (kennyt) [inline src/js/stores/__tests__/UnitHealthStore-test.js]: same here about that # of calls

File: src/js/stores/__tests__/NodeHealthStore-test.js (hunk 13 above)
Comment (daiyiwhale) [inline src/js/stores/__tests__/NodeHealthStore-test.js]: I was actually wondering about this, why are there 2 calls in the first place? Shouldn't there just be 1?

File: src/js/stores/NodeHealthStore.js (hunk 10 above)
Comment (daiyiwhale) [inline src/js/stores/NodeHealthStore.js]: you are totally right and really good at reading diff, thanksx10000

File: src/js/structs/NodesList.js (hunk 12 above)
Comment (daiyiwhale) [inline src/js/structs/NodesList.js]: hmm, like the last time this doesn't look like it needs the context. I'm going to remove it

File: src/js/stores/__tests__/NodeHealthStore-test.js (hunk 13 above)
Comment (kennyt) [inline src/js/stores/__tests__/NodeHealthStore-test.js]: just looked at the code. It should be two because when you `addChangeListener` and `shouldPoll` is true, it makes a call and in your `beforeEach` it instantly calls the success function. so that's call number 1.

call number 2 is when you do `handleServerAction` on line 70.

this test is passing because `shouldPoll` is incorrect atm

Comment (daiyiwhale) [inline src/js/stores/__tests__/NodeHealthStore-test.js]: ah ok, thanks!

File: src/js/structs/NodesList.js (hunk 12 above)
Comment (daiyiwhale) [inline src/js/structs/NodesList.js]: all good 👍

Comment (mlunoe) [inline src/js/structs/NodesList.js]: Hmmm... this works, but is used in a weird way in my opinion. We are setting the context to be `{ids: filters.ids}` and then `this.ids` refers to the `filters.ids`, but wouldn't it be more elegant to just refer to `filters.ids`, or even just create a variable `ids` outside of the callback?
I know this was not introduced by you, so we can do this in a separate PR, but I do think we need to change it

Comment (mlunoe) [inline src/js/structs/NodesList.js]: Actually, you made the same change here: https://github.com/dcos/dcos-ui/pull/7/files/3fdaa6d8e495df77e379fc8a141c7e01b9880dd4#diff-b99d7a3b11dae3abc53eb56ab919aa0bR25
So would you mind doing it here too?

File: src/js/utils/Units.js [hunk 15]
```diff
@@ -55,9 +54,7 @@ const Units = {
       return amount;
     }
 
-    options = _.extend({
-      decimalPlaces: 2
-    }, options);
+    options = Object.assign({decimalPlaces: 2}, options || {});
```
Comment (mlunoe) [inline src/js/utils/Units.js]: We can default assign `options` in the argument with `{}`, if we are only handling undefined here

File: src/js/structs/NodesList.js (hunk 12 above)
Comment (daiyiwhale) [inline src/js/structs/NodesList.js]: did it in this commit: https://github.com/dcos/dcos-ui/pull/7/commits/d1e422a2da5b905d7ccbe88696ab099fd97ed018

Comment (daiyiwhale) [inline src/js/structs/NodesList.js]: ....except I made a typo, one sec

Comment (daiyiwhale) [inline src/js/structs/NodesList.js]: alright, fixed now!

File: src/js/utils/Units.js (hunk 15 above)
Comment (daiyiwhale) [inline src/js/utils/Units.js]: Sorry, I don't quite understand this comment, could you rephrase?

Comment (mlunoe) [inline src/js/utils/Units.js]: Yeah, sorry. If we are only checking for `undefined` (and not `null` or other falsy values) when defaulting `options` to `{}` we could use default assign in the arguments of the function:
```js
contractNumber: function (amount, options = {}) {
```

This will only assign `options` if the argument is `undefined`, not if it `null` or other falsy values.

**Comment by mlunoe** [comment]
There is still an `underscore` reference here: https://github.com/dcos/dcos-ui/pull/7/files#diff-56274b172383b14152eb3c89445c36b3R1

File: src/js/utils/__tests__/Util-test.js [hunk 16]
```diff
@@ -5,6 +5,69 @@ var Util = require('../Util');
...
+      expect(last).toEqual(0);
+    });
+
+    it('should return null when given empty array', function () {
+      var array = [];
+      var last = Util.last(array);
+
+      expect(last).toEqual(null);
+    });
+
```
Comment (mlunoe) [inline src/js/utils/__tests__/Util-test.js]: Here is an extra newline, that's failing the build

**Comment by kennyt** [comment]
got a linting error (which is failing the test)

File: src/js/utils/__tests__/Util-test.js (hunk 16 above)
Comment (daiyiwhale) [inline src/js/utils/__tests__/Util-test.js]: Ah, there is it.

**Comment by daiyiwhale** [comment]
linting error addressed!
